### PR TITLE
test: Predicate parser on acceptance workload harness

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
## What changes are proposed in this pull request?

  - Rewrites the predicate parser to use bidirectional type checking for proper type inference, allowing  for predicates containing all the supported data types, including float, short, int, date, timestamp, etc. 
  - Adds support for arithmetic operators (+, -, *, /, %) and typed literals (DATE, TIMESTAMP,
  TIMESTAMP_NTZ)
  - Reorganizes acceptance test failure list with detailed root cause documentation
  - Adds resolve_column helper to schema for nested column path resolution

  Test plan
  - Existing predicate parser tests pass
  - New tests for arithmetic operators, typed literals, and edge cases
  - Acceptance workload tests pass (with documented expected failures)
  - cargo nextest run --workspace --all-features

### CODE REVIEW

File: acceptance/tests/acceptance_workloads_reader.rs [hunk 1]
```diff
@@ -304,91 +320,82 @@ const EXPECTED_KERNEL_FAILURES: &[(&str, &[&str])] = &[
         "variantShredding feature not supported",
         &["pv_002_upgrade_to_current/specs/pv_002_upgrade_to_current_read_latest"],
     ),
-    // Predicate evaluation fails due to incomplete type resolution, unsupported expressions,
```
Comment (OussamaSaoudi) [inline acceptance/tests/acceptance_workloads_reader.rs:303]: NOTE to reviewers: Deletion here means that the test is executing and passing.

Inclusion in these lists means that the test is executing, failing, and we expect the failure. In other words, it shows that there's work to do :)

File: benchmarks/src/predicate_parser.rs [hunk 2]
```diff
@@ -4,547 +4,879 @@
...
 //! - `IN (...)`, `NOT IN (...)`
 //! - `BETWEEN ... AND ...`
 //! - Column references and literal values (integers, floats, strings, booleans)
+//! - Typed literals: `DATE'...'`, `TIMESTAMP'...'`, `TIMESTAMP_NTZ'...'`
 //!
 //! Unsupported (returns error): `LIKE`, function calls (`HEX`, `size`, `length`),
-//! arithmetic expressions (`a % 100`), typed literals (`TIME '...'`).
+//! `TIME '...'` typed literals.
+//!
+//! # Type resolution
```
Comment (OussamaSaoudi) [inline benchmarks/src/predicate_parser.rs:17]: To reviewers: This is an intro to how the type checking works

File: acceptance/src/acceptance_workloads/workload.rs [hunk 3]
```diff
@@ -63,20 +63,22 @@ pub fn execute_read_workload(
...
     let table_schema = snapshot.schema();
 
-    // Build scan with column projection (no predicate pushdown - we filter after)
+    // Build scan with optional predicate and column projection
     let mut scan_builder = snapshot.scan_builder();
+
+    // Extract and parse the predicate if one is present
+    let predicate = if let Some(ref predicate_string) = read_spec.predicate {
+        let predicate = parse_predicate(predicate_string, &table_schema).map_err(Error::generic)?;
+        scan_builder = scan_builder.with_predicate(Arc::new(predicate.clone()));
```
Comment (DrakeLin) [inline acceptance/src/acceptance_workloads/workload.rs]: nit: make predicate an Arc before cloning

File: benchmarks/src/predicate_parser.rs [hunk 4]
```diff
@@ -1,524 +1,790 @@
...
 use delta_kernel::expressions::{
-    ArrayData, BinaryPredicateOp, ColumnName, Expression, Predicate, Scalar,
+    ArrayData, BinaryExpressionOp as KBinOp, BinaryPredicateOp as KPredOp, ColumnName,
+    Expression as KExpr, Predicate as KPred, Scalar,
+};
+use delta_kernel::schema::{ArrayType, DataType, PrimitiveType, Schema};
+use itertools::Itertools as _;
+use sqlparser::ast::{
+    BinaryOperator as PBinOp, Expr as PExpr, UnaryOperator as PUnaryOp, Value as PVal,
 };
```
Comment (sanujbasu) [inline benchmarks/src/predicate_parser.rs:54]: there is a subtle introduction of kernel predicates and sql parser predicates through prefixes here.
Could we add a comment specifically explaining that here?

File: benchmarks/src/predicate_parser.rs [hunk 5]
```diff
@@ -1,524 +1,790 @@
...
+    right: &PExpr,
+) -> Result<(KExpr, KExpr, DataType), Box<dyn std::error::Error>> {
+    let l_syn = synthesize_expr(schema, left);
+    let r_syn = synthesize_expr(schema, right);
+
+    let common_ty = match (&l_syn, &r_syn) {
+        (Some((_, l_ty)), Some((_, r_ty))) => find_common_type(l_ty, r_ty)
+            .ok_or_else(|| format!("Type mismatch: cannot compare {l_ty:?} with {r_ty:?}"))?,
+        (Some((_, ty)), None) | (None, Some((_, ty))) => ty.clone(),
+        (None, None) => {
```
Comment (sanujbasu) [inline benchmarks/src/predicate_parser.rs:292]: this arm has no test? Example : 5 > 3 with no column context

File: benchmarks/src/predicate_parser.rs [hunk 6]
```diff
@@ -1,524 +1,790 @@
...
+        .map(|e| -> Result<_, Box<dyn std::error::Error>> {
+            match check_expr(schema, &col_ty, e)? {
+                KExpr::Literal(s) => Ok(s),
+                _ => Err(format!("IN list elements must be literals, got: {e}").into()),
+            }
+        })
+        .try_collect()?;
+
+    // Check if any scalar is null to determine array nullability
+    let contains_null = scalars.iter().any(|s| s.is_null());
```
Comment (sanujbasu) [inline benchmarks/src/predicate_parser.rs:408]: do we have any test exercising this path?

File: benchmarks/src/predicate_parser.rs [hunk 7]
```diff
@@ -1,524 +1,790 @@
...
-) -> Result<Predicate, Box<dyn std::error::Error>> {
-    let col = convert_expr_to_expression(expr)?;
+/// Checks a literal value against an expected type and converts it to that type.
+/// Uses kernel's `PrimitiveType::parse_scalar` for parsing.
+fn check_literal(
+    expected_ty: &DataType,
+    value: &PVal,
+    is_negative: bool,
+) -> Result<KExpr, Box<dyn std::error::Error>> {
+    use PrimitiveType::*;
```
Comment (sanujbasu) [inline benchmarks/src/predicate_parser.rs:222]: includes in the top please

**Review by sanujbasu — APPROVED** [review]
Left some nits. Please address and merge
